controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,388 @@
+		logger.V(1).Info("pluginConfig is empty. Wasmplugin will be deleted if it exists")
+		utils.TagObjectToDelete(wasmPlugin)
+		return wasmPlugin, nil
+	}
+
+	pluginConfigStruct, err := pluginConfig.ToStruct()
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin.Spec.PluginConfig = pluginConfigStruct
+
+	// controller reference
+	if err := r.SetOwnerReference(gw, wasmPlugin); err != nil {
+		return nil, err
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) wasmPluginConfig(ctx context.Context, gw *gatewayapiv1.Gateway) (*wasm.Plugin, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin := &wasm.Plugin{
+		FailureMode:       wasm.FailureModeDeny,
+		RateLimitPolicies: make([]wasm.RateLimitPolicy, 0),
+	}
+
+	t, err := r.topologyIndexesFromGateway(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	rateLimitPolicies := t.PoliciesFromGateway(gw)
+
+	logger.V(1).Info("wasmPluginConfig", "#RLPS", len(rateLimitPolicies))
+
+	// Shallow copy by assignment
+	rateLimitPoliciesSorted := rateLimitPolicies
+
+	// Sort RLPs for consistent comparison with existing objects
+	sort.Sort(kuadrantgatewayapi.PolicyByCreationTimestamp(rateLimitPoliciesSorted))
+
+	for _, policy := range rateLimitPolicies {
+		rlp := policy.(*kuadrantv1beta2.RateLimitPolicy)
+		wasmRLP, err := r.WASMRateLimitPolicy(ctx, t, rlp, gw)
+		if err != nil {
+			return nil, err
+		}
+
+		if wasmRLP == nil {
+			// skip this RLP
+			continue
+		}
+
+		wasmPlugin.RateLimitPolicies = append(wasmPlugin.RateLimitPolicies, *wasmRLP)
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) topologyIndexesFromGateway(ctx context.Context, gw *gatewayapiv1.Gateway) (*kuadrantgatewayapi.TopologyIndexes, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	routeList := &gatewayapiv1.HTTPRouteList{}
+	// Get all the routes having the gateway as parent
+	err = r.Client().List(ctx, routeList, client.MatchingFields{HTTPRouteParents: client.ObjectKeyFromObject(gw).String()})
+	logger.V(1).Info("topologyIndexesFromGateway: list httproutes from gateway",
+		"gateway", client.ObjectKeyFromObject(gw),
+		"#HTTPRoutes", len(routeList.Items),
+		"err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	rlpList := &kuadrantv1beta2.RateLimitPolicyList{}
+	// Get all the rate limit policies
+	err = r.Client().List(ctx, rlpList)
+	logger.V(1).Info("topologyIndexesFromGateway: list rate limit policies",
+		"#RLPS", len(rlpList.Items),
+		"err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	policies := utils.Map(rlpList.Items, func(p kuadrantv1beta2.RateLimitPolicy) kuadrantgatewayapi.Policy { return &p })
+
+	t, err := kuadrantgatewayapi.NewTopology(
+		kuadrantgatewayapi.WithGateways([]*gatewayapiv1.Gateway{gw}),
+		kuadrantgatewayapi.WithRoutes(utils.Map(routeList.Items, ptr.To)),
+		kuadrantgatewayapi.WithPolicies(policies),
```

Comment:
```suggestion
	policies := utils.Map(rlpList.Items, func(p kuadrantv1beta2.RateLimitPolicy) kuadrantgatewayapi.Policy { return &p })
	routes := utils.Map(routeList.Items, func(r gatewayapiv1.HTTPRoute) *gatewayapiv1.HTTPRoute {return &r})

	t, err := kuadrantgatewayapi.NewTopology(
		kuadrantgatewayapi.WithGateways([]*gatewayapiv1.Gateway{gw}),
		kuadrantgatewayapi.WithRoutes(routes),
		kuadrantgatewayapi.WithPolicies(policies),
```
nit: Goland is highlighting this for me but it compiles and runs so not sure on this one .. 🤷 
<img width="827" alt="image" src="https://github.com/Kuadrant/kuadrant-operator/assets/24636860/7d4f472b-7b71-425a-8a10-4e0de5fa0794">

Reply:
`make run-lint` (golangci-lint) does not complain. It compiles. 

you sure it is valid lint issue? "cannot use generic function without instantiation" does not sound correct for this example.

Reply:
I am happy to change it to `routes := utils.Map(routeList.Items, func(r gatewayapiv1.HTTPRoute) *gatewayapiv1.HTTPRoute {return &r})` if you find this annoying. Let me know.

Reply:
This should fix it:

```suggestion
	policies := utils.Map(rlpList.Items, func(p kuadrantv1beta2.RateLimitPolicy) kuadrantgatewayapi.Policy { return &p })

	t, err := kuadrantgatewayapi.NewTopology(
		kuadrantgatewayapi.WithGateways([]*gatewayapiv1.Gateway{gw}),
		kuadrantgatewayapi.WithRoutes(utils.Map(routeList.Items, ptr.To[gatewayapiv1.HTTPRoute])),
		kuadrantgatewayapi.WithPolicies(policies),
```